Remove NetCDF groups (take 2)

joewallwork opened the pull request:
# Remove NetCDF groups (take 2)

Fixes #876
Supersedes #889

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR changes the format of the NetCDF input and output files used by nextSIM-DG so that they no longer use groups. The motivations for this are: (a) to be more consistent with how the community uses such files, and (b) to simplify integration with XIOS.

---
# Test Description

Input data generation scripts and CDL files will be updated for all tests
- [x] `core/test`
- [x] `physics/test`
- [x] `run`
- [x] `test`

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] This change conforms to the conventions described in the README

joewallwork committed:
Remove groups from source code

joewallwork committed:
Allow determining structure name

joewallwork committed:
Rework era5_test_data.py

joewallwork committed:
Rework topaz_test_data.py

joewallwork committed:
Provide structure name in XIOS test input

joewallwork committed:
Rework input files for ParaGrid test

joewallwork committed:
Fix accidentally dropped line in serial mode

joewallwork commented on core/src/Configurator.cpp:
```diff
@@ -62,7 +62,8 @@ void Configurator::setAdditionalConfiguration(AdditionalConfiguration* pAC) { ad
 {
     if (addConf()) {
-        addSStream(addConf()->read(source));
+        // FIXME: There's no longer any additional configuration to read?
+        // addSStream(addConf()->read(source));
```
Point for discussion - what's the purpose of the additional config?

joewallwork replied:
Addressed in adbe0d6cb4fa77d90bed36e259e7d75bb317ce1e.

joewallwork commented on core/src/NetcdfMetadataConfiguration.cpp:
```diff
@@ -20,17 +19,11 @@ std::stringstream NetcdfMetadataConfiguration::read(const std::string& source)
     std::stringstream config;
 
-    std::multimap<std::string, netCDF::NcVar> configs = configurationGroup.getVars();
+    // FIXME: It's reading coords but that has more than one entry!
+    std::multimap<std::string, netCDF::NcVar> configs = ncFile.getVars();
```
Note to self to revisit this FIXME.

joewallwork replied:
Addressed in adbe0d6cb4fa77d90bed36e259e7d75bb317ce1e.

joewallwork commented on core/src/ParaGridIO.cpp:
```diff
@@ -152,10 +149,9 @@ ModelState ParaGridIO::getModelState(const std::string& filePath)
             }
+            // Skip invalid dimension keys
             if (!dimensionKeys.count(dimKey)) {
-                throw std::out_of_range(
-                    std::string("No ModelArray::Type corresponds to the dimensional key ")
-                    + dimKey);
```
Note that I needed to rework this now that groups have been removed.

einola replied:
Is this note still relevant?

joewallwork replied:
I just thought I'd point out that the way these are processed had changed to the reviewer :)

joewallwork commented on core/src/ParallelNetcdfFile.cpp:
```diff
@@ -48,9 +48,9 @@ void NcFilePar::open_par(
     nullObject = false;
 }
 
-void netCDF::setVariableCollective(NcVar var, NcGroup group)
```
Note that pass by reference turned out to be required here to avoid crashing.

joewallwork commented on core/src/modules/include/IStructure.hpp:
```diff
@@ -79,13 +79,17 @@ class IStructure {
     // Node names in the default structure
 
     //! Returns the name of the metadata node.
+    // TODO: Is this still required after removing NetCDF groups?
     static const std::string metadataNodeName() { return "metadata"; }
     //! Returns the name of the data node.
+    // TODO: Is this still required after removing NetCDF groups?
     static const std::string dataNodeName() { return "data"; }
     //! The name of the group holding the definitive structure type
+    // TODO: Is this still required after removing NetCDF groups?
     static const std::string structureNodeName() { return "structure"; }
     //! The name of the node holding the name of the structure type processed
     //! by this class.
+    // TODO: Is this still required after removing NetCDF groups?
     static const std::string typeNodeName() { return "type"; }
```
Note to self to revisit: are all of these still required after this change?

joewallwork replied:
Addressed in https://github.com/nextsimhub/nextsimdg/pull/897/commits/8793520942a959d770a691090c9869672714f86b.

joewallwork commented on core/test/CMakeLists.txt:
```diff
@@ -14,6 +14,10 @@ execute_process(
     COMMAND "${Python_EXECUTABLE}" "${CMAKE_CURRENT_SOURCE_DIR}/old_names.py"
     WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
 )
+execute_process(
+    COMMAND "${Python_EXECUTABLE}" "${CMAKE_CURRENT_SOURCE_DIR}/ParaGridIO_input_test.py"
+    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
+)
```
I noticed that `ParaGridIO_input_test.nc` was included as a NetCDF file so dropped it and set things up so that it gets generated during the build.

joewallwork committed:
Do not write restart metadata for diagnostic files

joewallwork committed:
Fix structureNodeName usage

joewallwork committed:
Remove unused node names

joewallwork committed:
Remove groups from run directory

joewallwork commented:
[Rebased on top of `develop`.]

joewallwork committed:
Remove groups from test directory

joewallwork committed:
Don't read arrays from netcdf configuration

joewallwork commented on core/src/ParaGridIO.cpp:
```diff
@@ -350,28 +342,21 @@ void ParaGridIO::writeDiagnosticTime(
     if (isNew) {
         // Write the common structure and time metadata
         CommonRestartMetadata::writeStructureType(ncFile, meta);
-        CommonRestartMetadata::writeRestartMetadata(metaGroup, meta);
```
Note this was dropped because this section of code relates to diagnostics rather than restarts.

joewallwork replied:
Similarly in the XIOS variant.

joewallwork committed:
Use correct name for time_meta

joewallwork committed:
Drop writeRestartMetadata in consistent place

joewallwork reviewed (commented):
Okay I'm now happy with this. Ready for review!

joewallwork committed:
Merge branch 'develop' into issue876_remove-nc-groups-take2

einola commented on run/era5_topaz4_maker.py:
```diff
@@ -170,44 +171,38 @@ def topaz4_source_file_name(unix_time, path):
-    era_root = netCDF4.Dataset(era5_out_file, "w", format="NETCDF4")
-    structgrp = era_root.createGroup("structure")
-    structgrp.type = target_structure
+    era_ncFile = netCDF4.Dataset(era5_out_file, "w", format="NETCDF4")
```
```suggestion
    era5_ncFile = netCDF4.Dataset(era5_out_file, "w", format="NETCDF4")
```

Right? See also similar comment below.

joewallwork replied:
Thanks! I see you already fixed this.

einola commented on run/era5_topaz4_maker.py:
```diff
@@ -263,7 +258,7 @@ def topaz4_source_file_name(unix_time, path):
                 # Also use the windspeed loop to fill the time axis
                 nc_times[time_index] = unix_times_e[target_t_index]
-    era_root.close()
+    era_ncFile.close()
```
```suggestion
    era5_ncFile.close()
```

einola commented on run/era5_topaz4_maker.py:
```diff
@@ -170,44 +171,38 @@ def topaz4_source_file_name(unix_time, path):
     
     # Position and time variables
-    nc_lons = datagrp.createVariable("longitude", "f8", hfield_dims)
+    nc_lons = era5_ncFile.createVariable("longitude", "f8", hfield_dims)
```
When trying to run this script, it crashes here, saying

```
Writing ERA5 data to ERA5_2010-01-01_2010-01-02.nc
Traceback (most recent call last):
  File "/Users/einola/CLionProjects/nextsimdg/test/../run/era5_topaz4_maker.py", line 194, in <module>
    nc_lons[:, :] = element_lon
  File "src/netCDF4/_netCDF4.pyx", line 5487, in netCDF4._netCDF4.Variable.__setitem__
  File "src/netCDF4/_netCDF4.pyx", line 5770, in netCDF4._netCDF4.Variable._put
  File "src/netCDF4/_netCDF4.pyx", line 2014, in netCDF4._netCDF4._ensure_nc_success
RuntimeError: NetCDF: HDF error
```

I haven't been able to figure out how to fix this.

einola reviewed (changes requested):
Not an in-depth review. I had a quick look and they tried to check if everything worked as expected. Output and restart look fine, but ``era5_topaz4_maker.py`` doesn't run. I tried to debug it, but ran out of time, so I'll leave it for now. Maybe you'll see what the problem is.

einola committed:
Small bugfix

The netCDF file handle for the ERA5 output file was both being called
"era_ncFile" and "era5_ncFile". It's now always "era5_ncFile".

einola committed:
Fix to time handling

Time was not handled in a very good way any more, after removing the
groups (it was a bit weird with groups, but that doesn't matter now).
Now, we just have the "units" attribute for the "time" variable, and
that seems to work quite well.

einola commented:
I've pushed some fixes to ``era5_topaz4_maker.py``. It looks like it works now.

I made some changes to the way the time unit is handled, but I'm reasonably sure that the model assumes time is always given in seconds since 1970. So then it doesn't matter what I put in the ``units`` attribute for the time variable.

einola reviewed (approved):
Not a thorough review, rather I tested everything I could think of and now it works.
